### PR TITLE
reduce amount of lists that exist during runtime for custom materials

### PR DESCRIPTION
**Pull request by sohpeach**
## About The Pull Request
At the moment every object with custom materials has a unique list `obj.custom_materials` that represents the amount of each kind of material they contain. This PR caches these lists in the materials subsystem and re-uses them between objects that share the same materials.

Loading box station locally, this reduces the amount of lists in use by custom materials from over 13,000 to under 200. It's worth noting that this PR doesn't actually stop that many lists from being created (as the new code uses a temporary list), but it does reduce the amount of total lists the game is using.

I've had to edit some code that was modifying `custom_materials` on objects directly. Any code will be expected not to modify it too (this fact is documented.)

Something is probably broken, and it ain't the ORM.

## Why It's Good For The Game
Use less memory

## Changelog
:cl:
code: unique combinations of custom_materials lists are now shared between objects
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/materials.dm [hunk 1]
```diff
@@ -38,3 +41,21 @@ SUBSYSTEM_DEF(materials)
...
+		combo_params += "[path_name]=[materials_declaration[mat] * multiplier]"
+	sortTim(combo_params, /proc/cmp_text_asc)
+	var/combo_index = combo_params.Join("-")
+	var/list/combo = material_combos[combo_index]
+	if(!combo)
+		combo = list()
+		for(var/mat in materials_declaration)
+			combo[GetMaterialRef(mat)] = materials_declaration[mat] * multiplier
+		material_combos[combo_index] = combo
+	return combo
```
Comment (Rohesie) [inline code/controllers/subsystem/materials.dm:61]:
```suggestion
	return combo

```
Linter.

File: code/controllers/subsystem/materials.dm [hunk 2]
```diff
@@ -14,6 +14,8 @@ SUBSYSTEM_DEF(materials)
 	var/list/materials_by_category
 	///Dictionary of category || list of material types, mostly used by rnd machines like autolathes.
 	var/list/materialtypes_by_category
+	///A cache of all material combinations that have been used
+	var/list/material_combos
```
Comment (spookydonut) [inline code/controllers/subsystem/materials.dm:18]:
```suggestion
	var/list/list/material_combos
```

File: code/controllers/subsystem/materials.dm [hunk 3]
```diff
@@ -38,3 +41,22 @@ SUBSYSTEM_DEF(materials)
...
+	var/list/combo_params = list()
+	for(var/x in materials_declaration)
+		var/datum/material/mat = x
+		var/path_name = ispath(mat) ? "[mat]" : "[mat.type]"
+		combo_params += "[path_name]=[materials_declaration[mat] * multiplier]"
+	sortTim(combo_params, /proc/cmp_text_asc) // We have to sort now in case the declaration was not in order
+	var/combo_index = combo_params.Join("-")
+	var/list/combo = material_combos[combo_index]
+	if(!combo)
+		combo = list()
```
Comment (spookydonut) [inline code/controllers/subsystem/materials.dm:58]: this wont save this list to material_combos

**Review by spookydonut — CHANGES_REQUESTED** [review]

File: code/controllers/subsystem/materials.dm (hunk 3 above)
Comment (willox) [reply code/controllers/subsystem/materials.dm:58]: it's set on line 61

**Comment by actioninja** [comment]
be on hand if this breaks stuff, it looks fine in theory though.

**Comment by willox** [comment]
The biggest concern would be with any objects that modify the contents of their `custom_materials` list. I did go over every reference to it in the code and fixed the few that popped up, such as when resources from a stack are used.

**Pull request merged**